### ISSUE TITLE
Comma with no space not parsing as a separator

### ISSUE THREAD

**Issue by strickinato** [cortex-js/compute-engine#36]
I came across what _**appears**_ to be a bug in the parser related commas:

<img width="485" alt="image" src="https://user-images.githubusercontent.com/7566896/159590616-17090e07-fcc6-4ecf-ac9a-895ecbc59dac.png">

I would expect `1,2` (without a space) to still read as a `Sequence` or as multiple arguments, rather than collapsing into a single number.

(Perhaps there's a good reason for this I don't know about 😄)

**Version Info:**
Cortex: `"@cortex-js/compute-engine": "0.4.2"`
Node: v14.17.3

**Comment by arnog** [comment]
Hmmm... I can't seem to be able to reproduce this with the current version, `0.4.3`.

**Comment by strickinato** [comment]
🤔 -- I set about to play around with this to confirm with the new version, but ran into some trouble.



### 1. It seems like between 0.4.2 and 0.4.3, the api has changed significantly. Is that expected?

**0.4.2**
```
Welcome to Node.js v14.17.3.
> const computeEngine = require('@cortex-js/compute-engine');
> computeEngine
{
  ComputeEngine: [class In],
  LatexSyntax: [class si],
  evaluate: [Function (anonymous)],
  format: [Function (anonymous)],
  match: [Function: yi],
  parse: [Function (anonymous)],
  serialize: [Function (anonymous)],
  substitute: [Function: bi],
  version: '0.4.2'
}
> computeEngine.parse("1+2");
[ 'Add', { num: '1' }, { num: '2' } ]
```

**0.4.3**
Welcome to Node.js v14.17.3.
> const computeEngine = require('@cortex-js/compute-engine');
> computeEngine
{
  ComputeEngine: [class jt],
  DEBUG: true,
  DEFAULT_COMPLEXITY: 100000,
  version: '0.4.3'
}
> const c = new computeEngine.ComputeEngine()
undefined
> c.parse("1,2")
Mi {
  engine: <ref *1> jt {
    _cache: {},
    _commonSymbols: {
...
```

From what I gather poking around the source -  it seems that 0.4.3 now returns a `BoxedExpression` instead of an `Expression`). Is that right? Is there somewhere I can see the new API?

As of writing - the docs at https://cortexjs.io/compute-engine/ were updated in July 21 - so I'm guessing they're not the best way to explore the changes.


### 2. Trouble building 0.4.3

Something inside of rollup is not working for me - with a pretty unhelpful error. Is this something you've run into in your environment?
macOS 11.6.1
node 14.7.3

```
❯ npm run build
> @cortex-js/compute-engine@0.4.3 build /Users/aaronstrick/work/compute-engine
> bash ./scripts/build.sh

 Compute Engine   ✔ Cleaning output directories
 Compute Engine   ✔  Building TypeScript declaration files
 Compute Engine   羽 Making a production build
[!] SyntaxError: Unexpected identifier
SyntaxError: Unexpected identifier
    at Loader.moduleStrategy (internal/modules/esm/translators.js:145:18)
    at link (internal/modules/esm/module_job.js:63:21)
```
Running the alternative commands for rollup that you have commented out in the `scripts/build.sh` give the same.



_I acknowledge that these don't really have to do with this particular ticket, but I'm not sure where else to log them, because they may be specific to me trying to see if this is fixed in 0.4.3 - but they may be relevant to the new release as a whole 😄  -- Thanks so much for your help!_

**Comment by arnog** [comment]
0.4.3 is available in npm and via a CDN. Depending on what you're trying to do this may be able to work around the build error you're getting. 

I'm not sure about those build errors, but it could be a mismatch of node version. I am currently using node 17.3.1, which has support for importing JSON modules, and I have a hunch that may be what you're running into.

I have updated the doc on cortex.js.io, but yes, the API has changed. You now only use a `ComputeEngine` instance, and its `.parse()` or `.box()` methods. The result of `.parse()` is indeed a boxed expression, i.e. a MathJSON expression wrapped in a JS object. To get access to the "raw" MathJSON, use the `.json` property.

Where before you would have `console.log(parse("1+1"))` you would now use `console.log(ce.parse("1+1").json)`. You can also get a serialized LaTeX using the `.latex` property. To manipulate the expression (i.e. simplify it, evaluate it, etc...) use the corresponding methods on  the expression object:

```ts
const expr = ce.parse("1+1");
console.log(expr.simplify().json);
console.log(expr.evaluate().latex);
```

**Comment by strickinato** [comment]
> I'm not sure about those build errors, but it could be a mismatch of node version. I am currently using node 17.3.1

This was totally correct. node 17 got me going!

> 0.4.3 is available in npm and via a CDN. Depending on what you're trying to do this may be able to work around the build error you're getting.

😄 Yup! Planning on updating my main application to 0.4.3. However, when I ran into the unexpected API changes, I reached for building from source so I could play with it separately and figure out what's going on.

Also, there are a few open issues that we'd really like to address (eg. https://github.com/cortex-js/compute-engine/issues/32), so I wanted to be able to build from source to start trying to make some PRS.

It looks like, from initial playing, that some/many of them _**have been fixed**_!!! Yay!! Thank you so much!

I'm going to spend today:
1. Updating our app
2. Testing out some more of those issues against 0.4.3 and see about closing some of them 😄 

**Re: docs**
Would it be helpful if I took a little time to change the examples to reflect the new API? Let me know

**Comment by arnog** [comment]
> This was totally correct. node 17 got me going!

OK, great. I'm not sure exactly what the minimum version is. Might be lower than 17. The node ecosystem is changing quite a bit with regard to module support and that's creating some unpleasant turbulence.

> > 0.4.3 is available in npm and via a CDN. Depending on what you're trying to do this may be able to work around the build error you're getting.
> 
> 😄 Yup! Planning on updating my main application to 0.4.3. However, when I ran into the unexpected API changes, I reached for building from source so I could play with it separately and figure out what's going on.

Sounds good.

> Also, there are a few open issues that we'd really like to address (eg. #32), so I wanted to be able to build from source to start trying to make some PRS.

I think a number of issues have been addressed with this new version. Some parsing patterns that failed before , or that could not be handled correctly, particularly with nested delimiters, work correctly now. 

In general, as you've noticed, this version is a big change, both in terms of the API surface, but even more so, in terms of the architecture of the implementation. It was painful to go through, but it was necessary and there is now a much better foundation that will allow iterating and making some progress much faster now. 

I'm looking forward to any PR you might be interested in sharing. Don't hesitate to let me know ahead of time either if you're considering working in a specific area. I will also be pushing updates much more frequently now :)


> It looks like, from initial playing, that some/many of them _**have been fixed**_!!! Yay!! Thank you so much!

Yes, I haven't had the time yet to go through the open issues and check their status, but if you do see some that are now fixed, please feel free to close them or to add a comment indicating their status, that would be very helpful.

> **Re: docs** Would it be helpful if I took a little time to change the examples to reflect the new API? Let me know

Yes, that would be great. I still plan to do another pass to streamline the doc. I'm also aware that the SDK Reference is currently broken and I will fix this as well in the next couple of days.

**Comment by strickinato** [comment]
> Yes, I haven't had the time yet to go through the open issues and check their status
I made some updates!

> Don't hesitate to let me know ahead of time either if you're considering working in a specific area
Still in the process of updating this in our app, and sorting through our existing issues / the issues are problematic for us and come from within cortex -- but whatever comes at the center of that will be "the area in which I focus" -- and will let you know before starting work 😄

**Comment by arnog** [comment]
Thank you!

I've put together a transition guide which you might find useful: https://github.com/cortex-js/compute-engine/blob/main/CHANGELOG.md

**Comment by strickinato** [comment]
Super helpful! Much appreciated 🙏 

I haven't had a chance to get all the way through it, because I'm actually stuck on another issue:

We are still using webpack 4, which [does not have support for optional chaining](https://github.com/webpack/webpack/issues/10227#issuecomment-642734920). I'm still struggling to try to successfully build the new version into our bundle.

I don't know if others are running into this, but I wonder if it would be fairly quick / helpful to a wider audience to either get rid of the optional chain or have the rollup config spit out compatible js. (Perhaps this should go in another issue 🤔 )

**Comment by arnog** [comment]
Sigh... yes, I feel for you. It's really sad when *all* evergreen browsers natively support a feature, but you can't use it because of an outdated component in your toolchain...

I could switch to target ECMA2017, but it's a 6% increase in minified code size.

I suppose I could add another variant to the build output targeting "vintage" environments. You don't use the `.esm.js` variant, do you?

**Comment by arnog** [comment]
...  or I could have the non-module build targets build for ES5, and the module targets build for ES2020...

**Comment by arnog** [comment]
OK, I've checked in a change: the non-module targets (i.e. `compute-engine.min.js`) are now built to target "vintage" environments (i.e. don't include optional chaining or other "modern" features). The module variants (i.e. `compute-engine.min.esm.js`) target ES2020.

Would you need a version pushed to npm or can you try it as is?

**Comment by strickinato** [comment]
Thank you for all your help! Some notes for anyone who encounters this in the future 🤷 

## w/r/t to the actual issue:

Solved!
![image](https://user-images.githubusercontent.com/7566896/161165025-b94cd629-2a65-4dd1-90a4-58df96b89fc8.png)

⚠️ Note that it is "Delimiter" now and not "Parentheses" --  `ce.parse('(3,3)')` is `[ 'Delimiter', 3, 3 ]`, not `[ 'Parentheses', 3, 3 ]`

### To succesfully import "vintage" version:

 I used the [webpack alias feature](https://v4.webpack.js.org/configuration/resolve/#resolvealias) to directly specify the target build.

```js
// note the $ for an exact match, to avoid recursive resolution issues
alias: {
    '@cortex-js/compute-engine$': '@cortex-js/compute-engine/dist/compute-engine.min.js',
    '@cortex-js/math-json$': '@cortex-js/compute-engine/dist/math-json.min.js',
}
```

To include the typescript type definitions, a similar thing needs to happen in the `.tsconfig`

```json
{
  "compilerOptions": {
    "paths": {
       "@cortex-js/*": ["node_modules/@cortex-js/compute-engine/dist/types/*"]
      }
   }
}
```

**Comment by arnog** [comment]
👍 Great! And thanks for the update.
